README.md
# Consulta Processual

Este projeto consiste em uma aplicação de consulta processual com front-end construído usando Next.js (ou React) e back-end construído usando Node.js e Express.

## Estrutura do projeto

O projeto é dividido em duas partes principais:

- `consulta-processual-frontend`: Aplicação front-end criada com Next.js (ou React).
- `consulta-processual-backend`: Aplicação back-end criada com Node.js e Express.

## Como executar o projeto

### Front-end

1. Navegue até a pasta `consulta-processual-frontend`.
2. Instale as dependências com `npm install` ou `yarn`.
3. Inicie o servidor de desenvolvimento com `npm run dev` ou `yarn dev`.

### Back-end

1. Navegue até a pasta `consulta-processual-backend`.
2. Instale as dependências com `npm install` ou `yarn`.
3. Inicie o servidor com `node index.js`.

## Configuração do banco de dados

1. Crie um banco de dados PostgreSQL e configure as variáveis de ambiente no arquivo .env na pasta consulta-processual-backend com as informações de conexão. Exemplo de arquivo .env:

DB_HOST=localhost
DB_PORT=5432
DB_USER=myuser
DB_PASSWORD=mypassword
DB_NAME=mydatabase

2. Instale a biblioteca dotenv no back-end para carregar as variáveis de ambiente. Você pode instalá-la executando npm install dotenv ou yarn add dotenv.
3. Importe a biblioteca dotenv no início do arquivo index.js e configure as variáveis de ambiente. Adicione a seguinte linha no topo do arquivo:

```javascript
require("dotenv").config();
```

4. Atualize o arquivo database.js para usar as variáveis de ambiente em vez de valores fixos:

```javascript
const { Sequelize } = require("sequelize");

const sequelize = new Sequelize(
  process.env.DB_NAME,
  process.env.DB_USER,
  process.env.DB_PASSWORD,
  {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    dialect: "postgres",
    logging: false,
  },
);
```

module.exports = sequelize;

5. Certifique-se de que seu banco de dados esteja em execução e que as informações de conexão estejam corretas.

## Acesso à aplicação

Após iniciar os servidores, acesse o front-end em `http://localhost:3000` e o back-end em `http://localhost:3001`.


HISTORY.md
# Histórico do Projeto

## Decisão

- Escolha do front-end: Optei por utilizar o Next.js (ou React) no front-end devido à sua facilidade de uso, popularidade e recursos avançados, como SSR e SSG.
- Escolha do back-end: Decidi usar Node.js e Express no back-end por serem simples, rápidos e permitirem o uso de JavaScript em todo o projeto.
- Escolha do banco de dados: Decidi usar o PostgreSQL como banco de dados por ser um sistema gerenciador de banco de dados open-source, robusto e com bom suporte à linguagem SQL.

## Passos realizados

1. Configuração inicial do ambiente de desenvolvimento com Next.js (ou React) e Node.js + Express.
2. Verificação da conexão com o banco de dados PostgreSQL.
3. Criação do arquivo `database.js` para gerenciar a conexão com o banco de dados.
4. Criação do arquivo `index.js` para configurar um servidor básico com Express.
5. Definição do modelo "Processo" e sincronização com o banco de dados.
6. Implementação das rotas e controladores de movimentação, incluindo a lógica de criação, atualização, exclusão e consulta de movimentações.
7. Adição e configuração do chance.js para geração de dados aleatórios durante o desenvolvimento e testes.
8. Implementação de validação de entrada de dados e tratamento de erros nos controladores.

## Ideias e recursos futuros

- Implementar a lógica completa de consulta e exibição de processos no front-end e back-end.
- Adicionar suporte a testes no front-end e back-end.
- Configurar o banco de dados e modelos necessários para armazenar informações sobre os processos e movimentações.
- Adicionar autenticação e autorização para proteger as informações dos processos.

## Arquiteturas testadas

- Neste momento, estou trabalhando com uma arquitetura que divide o projeto em front-end e back-end separados, ambos escritos em JavaScript (usando Next.js ou React no front-end e Node.js + Express no back-end).


consulta-processual-backend/seed.js
const Chance = require('chance');
const chance = new Chance();

const Processo = require('./models/Processo');
const Movimentacao = require('./models/Movimentacao');

// Função que gera um CNJ aleatório
const generateCnj = () => {
  const numbers = [...Array(20)].map(() => Math.floor(Math.random() * 10));
  return numbers.join('');
}

// Função que popula o banco de dados com dados fictícios
const seedDatabase = async () => {
  // Cria 10 processos fictícios
  const processos = await Processo.bulkCreate([...Array(10)].map(() => ({
    cnj: generateCnj(),
    partes: chance.name(),
    tribunal: chance.company(),
    dataInicio: chance.date({ year: 2022 }),
  })));

  // Para cada processo, cria um número aleatório de movimentações
  for (const processo of processos) {
    const numMovimentacoes = chance.integer({ min: 1, max: 10 });

    await Movimentacao.bulkCreate([...Array(numMovimentacoes)].map(() => ({
      descricao: chance.sentence(),
      data: chance.date({
        min: processo.dataInicio,
        max: new Date()
      }),
      processoId: processo.id,
    })));
  }

  console.log('Dados populados com sucesso.');
}

seedDatabase();

consulta-processual-backend/index.js
const express = require('express');
const cors = require('cors');
const sequelize = require('./database');
const Processo = require('./models/Processo');
const Movimentacao = require('./models/Movimentacao');
const processoController = require('./controllers/processoController');
const movimentacaoController = require('./controllers/movimentacaoController');

const app = express();
const PORT = process.env.PORT || 3000;

app.use(express.json());
app.use(cors());

app.post('/processos', processoController.validateProcesso, processoController.createProcesso);
app.get('/processos', processoController.getProcessos);
app.get('/processos/:id', processoController.getProcesso);
app.put('/processos/:id', processoController.validateProcesso, processoController.updateProcesso);
app.delete('/processos/:id', processoController.deleteProcesso);

app.post('/movimentacoes', movimentacaoController.validateMovimentacao, movimentacaoController.createMovimentacao);
app.get('/movimentacoes', movimentacaoController.getMovimentacoes);
app.get('/movimentacoes/:id', movimentacaoController.getMovimentacao);
app.put('/movimentacoes/:id', movimentacaoController.validateMovimentacao, movimentacaoController.updateMovimentacao);
app.delete('/movimentacoes/:id', movimentacaoController.deleteMovimentacao);

app.get('/', (req, res) => {
  res.send('Hello World');
});

app.use((req, res, next) => {
  res.status(404).json({ message: 'Rota não encontrada.' });
});

app.use((err, req, res, next) => {
  console.error(err);
  res.status(500).json({ message: 'Ocorreu um erro interno no servidor.' });
});

app.listen(PORT, () => {
  console.log(`Server is running on port ${PORT}`);
});

(async () => {
  try {
    await sequelize.authenticate();
    console.log('Connection to the database has been established successfully.');

    await sequelize.sync({ force: true });
    console.log('Database tables have been created and synchronized.');
  } catch (error) {
    console.error('Unable to connect to the database:', error);
  }
})();

consulta-processual-backend/controllers/movimentacaoController.js
const Movimentacao = require('../models/Movimentacao');
const Processo = require('../models/Processo');
const { body, validationResult } = require('express-validator');

exports.validateMovimentacao = [
  body('descricao').notEmpty().withMessage('O campo descrição é obrigatório.'),
  body('dataMovimentacao').isDate().withMessage('A data de movimentação deve ser uma data válida.'),
];

exports.createMovimentacao = async (req, res) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  try {
    const { descricao, dataMovimentacao } = req.body;
    const processoId = req.params.id;
    const processo = await Processo.findByPk(processoId);

    if (!processo) {
      return res.status(404).json({ message: 'Processo não encontrado.' });
    }

    const movimentacao = await Movimentacao.create({ descricao, dataMovimentacao, processoId });

    return res.status(201).json(movimentacao);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao criar movimentação.' });
  }
};

exports.getMovimentacoes = async (req, res) => {
  try {
    const processoId = req.params.id;
    const processo = await Processo.findByPk(processoId);

    if (!processo) {
      return res.status(404).json({ message: 'Processo não encontrado.' });
    }

    const movimentacoes = await Movimentacao.findAll({ where: { processoId } });
    res.status(200).json(movimentacoes);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao buscar movimentações.' });
  }
};

exports.getMovimentacao = async (req, res) => {
  try {
    const movimentacaoId = req.params.id;
    const movimentacao = await Movimentacao.findByPk(movimentacaoId);

    if (!movimentacao) {
      return res.status(404).json({ message: 'Movimentação não encontrada.' });
    }

    res.status(200).json(movimentacao);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao buscar movimentação.' });
  }
};

exports.updateMovimentacao = async (req, res) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  try {
    const movimentacaoId = req.params.id;
    const movimentacaoData = req.body;

    const movimentacao = await Movimentacao.findByPk(movimentacaoId);

    if (!movimentacao) {
      return res.status(404).json({ message: 'Movimentação não encontrada.' });
    }

    await movimentacao.update(movimentacaoData);

    res.status(200).json(movimentacao);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao atualizar movimentação.' });
  }
};

exports.deleteMovimentacao = async (req, res) => {
  try {
    const movimentacaoId = req.params.id;
    const movimentacao = await Movimentacao.findByPk(movimentacaoId);

    if (!movimentacao) {
      return res.status(404).json({ message: 'Movimentação não encontrada.' });
    }

    await movimentacao.destroy();
    res.status(204).json({ message: 'Movimentação excluída com sucesso.' });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao excluir movimentação.' });
  }
};

consulta-processual-backend/controllers/processoController.js
const Processo = require('../models/Processo');
const { body, validationResult } = require('express-validator');

exports.validateProcesso = [
  body('cnj').notEmpty().withMessage('O campo CNJ é obrigatório.'),
  body('partes').notEmpty().withMessage('O campo partes é obrigatório.'),
  body('tribunal').notEmpty().withMessage('O campo tribunal é obrigatório.'),
  body('dataInicio').isDate().withMessage('A data de início deve ser uma data válida.'),
];

exports.createProcesso = async (req, res) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  try {
    const { cnj, partes, tribunal, dataInicio } = req.body;
    const processo = await Processo.create({ cnj, partes, tribunal, dataInicio });

    return res.status(201).json(processo);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao criar processo.' });
  }
};

exports.updateProcesso = async (req, res) => {
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).json({ errors: errors.array() });
  }

  try {
    const processoId = req.params.id;
    const processoData = req.body;

    const processo = await Processo.findByPk(processoId);

    if (!processo) {
      return res.status(404).json({ message: 'Processo não encontrado.' });
    }

    await processo.update(processoData);

    res.status(200).json(processo);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao atualizar processo.' });
  }
};

exports.getProcessos = async (req, res) => {
  try {
    const processos = await Processo.findAll();
    res.status(200).json(processos);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao buscar processos.' });
  }
};

exports.getProcesso = async (req, res) => {
  try {
    const processoId = req.params.id;
    const processo = await Processo.findByPk(processoId);

    if (!processo) {
      return res.status(404).json({ message: 'Processo não encontrado.' });
    }

    res.status(200).json(processo);
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao buscar processo.' });
  }
};

exports.deleteProcesso = async (req, res) => {
  try {
    const processoId = req.params.id;
    const processo = await Processo.findByPk(processoId);

    if (!processo) {
      return res.status(404).json({ message: 'Processo não encontrado.' });
    }

    await processo.destroy();
    res.status(204).json({ message: 'Processo excluído com sucesso.' });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: 'Erro ao excluir processo.' });
  }
};